Thanks for the patch, I kicked off some builds, will report back.

The `x86_64-darwin` build still fails very early:


```
@nix { "action": "setPhase", "phase": "buildPhase" }
building
build flags: -j8
[0/2] Re-checking globbed directories...
[1/10015] Building CXX object contrib/libunwind-cmake/CMakeFiles/unwind.dir/__/libunwind/src/Unwind-EHABI.cpp.o
[2/10015] Building CXX object contrib/libunwind-cmake/CMakeFiles/unwind.dir/__/libunwind/src/Unwind-seh.cpp.o
FAILED: contrib/libunwind-cmake/CMakeFiles/unwind.dir/__/libunwind/src/Unwind-seh.cpp.o 
/nix/store/dx25g7vqky7bp48q2v567wr4qgrpdqmf-clang-wrapper-16.0.6/bin/clang++ -D_DEBUG -D_LIBUNWIND_IS_NATIVE_ONLY -D_LIBUNWIND_NO_HEAP=1 -isystem /tmp/nix-build-clickhouse-23.10.3.5.drv-0/source/contrib/libunwind/include -fdiagnostics-color=always -fsized-deallocation  -mpopcnt -fasynchronous-unwind-tables -ffile-prefix-map=/tmp/nix-build-clickhouse-23.10.3.5.drv-0/source=. -falign-functions=32 -mbranches-within-32B-boundaries  -stdlib=libc++ -fdiagnostics-absolute-paths -fstrict-vtable-pointers -O3 -DNDEBUG -std=c++2b -mmacosx-version-min=10.15   -D OS_DARWIN -O3 -fno-exceptions -funwind-tables -fno-sanitize=all -nostdinc++ -fno-rtti -MD -MT contrib/libunwind-cmake/CMakeFiles/unwind.dir/__/libunwind/src/Unwind-seh.cpp.o -MF contrib/libunwind-cmake/CMakeFiles/unwind.dir/__/libunwind/src/Unwind-seh.cpp.o.d -o contrib/libunwind-cmake/CMakeFiles/unwind.dir/__/libunwind/src/Unwind-seh.cpp.o -c /tmp/nix-build-clickhouse-23.10.3.5.drv-0/source/contrib/libunwind/src/Unwind-seh.cpp
In file included from /tmp/nix-build-clickhouse-23.10.3.5.drv-0/source/contrib/libunwind/src/Unwind-seh.cpp:13:
In file included from /tmp/nix-build-clickhouse-23.10.3.5.drv-0/source/contrib/libunwind/src/config.h:16:
In file included from /nix/store/jh0ibzj56bmw6vd9g67jc0x7cv8wmm7m-libSystem-11.0.0/include/assert.h:44:
In file included from /nix/store/jh0ibzj56bmw6vd9g67jc0x7cv8wmm7m-libSystem-11.0.0/include/stdlib.h:128:
/nix/store/jh0ibzj56bmw6vd9g67jc0x7cv8wmm7m-libSystem-11.0.0/include/malloc/_malloc.h:40:48: error: expected function body after function declarator
void    *malloc(size_t __size) __result_use_check __alloc_size(1);
                                                  ^
/nix/store/jh0ibzj56bmw6vd9g67jc0x7cv8wmm7m-libSystem-11.0.0/include/malloc/_malloc.h:41:64: error: expected function body after function declarator
void    *calloc(size_t __count, size_t __size) __result_use_check __alloc_size(1,2);
                                                                  ^
/nix/store/jh0ibzj56bmw6vd9g67jc0x7cv8wmm7m-libSystem-11.0.0/include/malloc/_malloc.h:43:62: error: expected function body after function declarator
void    *realloc(void *__ptr, size_t __size) __result_use_check __alloc_size(2);
                                                                ^
/nix/store/jh0ibzj56bmw6vd9g67jc0x7cv8wmm7m-libSystem-11.0.0/include/malloc/_malloc.h:45:22: error: expected function body after function declarator
void    *valloc(size_t) __alloc_size(1);
                        ^
/nix/store/jh0ibzj56bmw6vd9g67jc0x7cv8wmm7m-libSystem-11.0.0/include/malloc/_malloc.h:50:78: error: expected function body after function declarator
void    *aligned_alloc(size_t __alignment, size_t __size) __result_use_check __alloc_size(2) __OSX_AVAILABLE(10.15) __IOS_AVAILABLE(13.0) __TVOS_AVAILABLE(13.0) __WATCHOS_AVAILABLE(6.0);
                                                                             ^
```

I uploaded the full build log [here](https://gist.github.com/flokli/bc08b59653f2024220298ee46303900c).

--- a/pkgs/servers/clickhouse/default.nix
+++ b/pkgs/servers/clickhouse/default.nix
@@ -12,7 +12,6 @@
 , darwin
 , findutils
 
-# currently for BLAKE3 hash function
 , rustSupport ? true
 
 , corrosion

--- a/pkgs/servers/clickhouse/default.nix
+++ b/pkgs/servers/clickhouse/default.nix
@@ -152,6 +152,10 @@ in mkDerivation rec {
     popd
 
     cargoSetupPostPatchHook() { true; }
+  '' + lib.optionalString stdenv.isDarwin ''
+    # Make sure Darwin invokes lld.ld64 not lld.
+    substituteInPlace cmake/tools.cmake \
+      --replace '--ld-path=''${LLD_PATH}' '-fuse-ld=lld'
   '';
 
   cmakeFlags = [
@@ -159,6 +163,12 @@ in mkDerivation rec {
     "-DCOMPILER_CACHE=disabled"
     "-DENABLE_EMBEDDED_COMPILER=ON"
   ];
+
+  env = lib.optionalAttrs stdenv.isDarwin {
+    # Silence ``-Wimplicit-const-int-float-conversion` error in MemoryTracker.cpp and
+    # ``-Wno-unneeded-internal-declaration` TreeOptimizer.cpp.
+    NIX_CFLAGS_COMPILE = "-Wno-implicit-const-int-float-conversion -Wno-unneeded-internal-declaration";
+  };
 
   # https://github.com/ClickHouse/ClickHouse/issues/49988
   hardeningDisable = [ "fortify" ];

